src/theory/strings/theory_strings.cpp

```diff
@@ -258,6 +248,39 @@ Node TheoryStrings::getLength( Node t, std::vector< Node >& exp ) {
   return getLengthExp( t, exp, t );
 }
 
+Node TheoryStrings::getNormalString(Node x, std::vector<Node>& nf_exp)
+{
+  if (!x.isConst())
+  {
+    Node xr = getRepresentative(x);
+    if (d_normal_forms.find(xr) != d_normal_forms.end())
+    {
+      Node ret = mkConcat(d_normal_forms[xr]);
+      nf_exp.insert(nf_exp.end(),
+                    d_normal_forms_exp[xr].begin(),
+                    d_normal_forms_exp[xr].end());
+      addToExplanation(x, d_normal_forms_base[xr], nf_exp);
+      Trace("strings-debug")
+          << "Term: " << x << " has a normal form " << ret << std::endl;
+      return ret;
+    }
+    else
+    {
+      if (x.getKind() == kind::STRING_CONCAT)
+      {
+        std::vector<Node> vec_nodes;
+        for (unsigned i = 0; i < x.getNumChildren(); i++)
+        {
+          Node nc = getNormalString(x[i], nf_exp);
+          vec_nodes.push_back(nc);
+        }
+        return mkConcat(vec_nodes);
+      }
+    }
+  }
+  return x;
```

Comment:
Why are we returning `x` here and not the representative of `x`? (Just curious, the code is the same as in the original)

Reply:
I think the two are equivalent. If x has a representative that is distinct from x, then it exists in the equality engine, and hence should have a normal form. I'll add an assertion for this.